contracts/voting-verifier/src/contract/execute.rs

```diff
@@ -157,6 +168,117 @@ pub fn verify_messages(
+) -> Result<Response, ContractError> {
+    if messages_with_payloads.is_empty() {
+        return Err(report!(ContractError::EmptyMessages));
+    }
+
+    let config = CONFIG.load(deps.storage).expect("failed to load config");
+
+    let messages = messages_with_payloads.try_map(|message_with_payload| {
+        // Message needs to be to ITS Hub
+        if message_with_payload.message.destination_chain.as_ref()
+            != config.axelar_chain_name.as_ref()
+            || message_with_payload.message.destination_address.as_str()
+                != config.its_hub_address.as_str()
+        {
+            return Err(ContractError::InvalidMessage.into());
+        }
+
+        // Payload needs to be correct, corresponding to the payload hash
+        if message_with_payload.message.payload_hash.as_slice()
+            != Keccak256::digest(&message_with_payload.payload).as_slice()
+        {
+            return Err(ContractError::InvalidPayload.into());
+        }
+
+        validate_source_chain(message_with_payload.message, &config.source_chain)
+            .and_then(|message| validate_source_address(message, &config.address_format))
+            .and_then(|message| {
+                message_status(deps.as_ref(), &message, env.block.height)
+                    .map(|status| (status, message, message_with_payload.payload))
+            })
+    })?;
+
+    let msgs_to_verify_with_payloads: Vec<(Message, HexBinary)> = messages
+        .into_iter()
+        .filter_map(|(status, message, payload)| match status {
+            VerificationStatus::NotFoundOnSourceChain
+            | VerificationStatus::FailedToVerify
+            | VerificationStatus::Unknown => Some((message, payload)),
+            VerificationStatus::InProgress
+            | VerificationStatus::SucceededOnSourceChain
+            | VerificationStatus::FailedOnSourceChain => None,
+        })
+        .collect();
+
+    if msgs_to_verify_with_payloads.is_empty() {
+        return Ok(Response::new());
+    }
+
+    // let message = msg_to_verify.unwrap();
+
+    let snapshot = take_snapshot(deps.as_ref(), &config.source_chain)?;
+    let participants = snapshot.participants();
+    let expires_at = calculate_expiration(env.block.height, config.block_expiry.into())?;
+
+    let id = create_messages_poll(deps.storage, expires_at, snapshot, 1)?;
+
+    for (idx, (message, _)) in msgs_to_verify_with_payloads.iter().enumerate() {
+        poll_messages()
+            .save(
+                deps.storage,
+                &message.hash(),
+                &state::PollContent::<Message>::new(message.clone(), id, idx),
+            )
+            .change_context(ContractError::StorageError)?;
+    }
+
+    let stacks_abi_transformer: stacks_abi_transformer::Client =
+        client::ContractClient::new(deps.querier, &config.stacks_abi_transformer).into();
+
+    let mut clarity_payloads = Vec::new();
+    let messages = msgs_to_verify_with_payloads
```

Comment:
to avoid mut, you could just create a vector of tuples and then unzip

Reply:
Done